#!/bin/bash

## ---------------------------------------

## this script processes R2 FASTQ files to analyze allele sequences and determine the most likely typed alleles

## main input: 
## - R2 FASTQ files from a specified directory (referred to in this script as fastq="$1")
## - a sequence file containing HLA headers and sequences of interest
##   generated by the previous rule (extract_sequences) (referred to in this script as seq="$2)

## for each FASTQ file, sequences from the input file are matched against the reads in the R2 FASTQ file
## the script counts how many times each sequence appears in the file
## the term "allele" here refers to a sequence variant that occurs frequently (in high percentage)
## the top 2 most frequent sequences are identified, and their occurrence percentages are calculated
## if the difference between the percentages of the top 2 alleles is greater than 50%
## the higher percentage allele is classified as the "typed" allele, suggesting there's only one dominant allele
## otherwise, both alleles are considered, suggesting there are two alleles

## output file looks something similar to the following: 
## >HLA-A|NM_002116.7|Reference_end : CTGTGTTCGTGTAGG
## Top 2 alleles:
## 2156   CTGTGTTCGTGTAGGCATAAAGTGAGGAGGTGGGGAGACCACCCCACCCCCATGTCCACCATGACCCTCTTCCC
## 195566 CTGTGTTCGTGTAGGCATAATGTGAGGAGGTGGGGAGACCACCCCACCCCCATGTCCACCATGACCCTCTTCCC
## Total number of reads: 230991
## Percentage of the top 2 alleles:
## 0.93%
## 84.66%
## Difference between the top 2 alleles: 83.73%
## Typed alleles:
## 195566 CTGTGTTCGTGTAGGCATAATGTGAGGAGGTGGGGAGACCACCCCACCCCCATGTCCACCATGACCCTCTTCCC

## ---------------------------------------

## input data
fastq="$1"
seq="$2"
num_processes="$3"
output_dir="$4"

ls "$fastq"/*_R2.fastq.gz | xargs -I {} -P "$num_processes" sh -c '
    file="{}"
    output_dir="$1"
    seq="$2"
    
    output_file="$output_dir/alleles_$(basename {} _R2.fastq.gz).txt"
    while IFS= read -r header; do

        read -r sequence
        match=$(zcat "$file" | grep "^$sequence")
        sum_counts=$(echo "$match" | wc -l)
        
        ## this command processes each line of the R2 file, matching reads against a specific HLA sequence
        ## given that each read spans four lines, it extracts only the relevant sequence information
        ## all sequences are concatenated, eliminating any empty lines between them
        ## finally, it sorts the sequences and counts the unique occurrences of each specific sequence
        allele_counts=$(echo "$match" | grep -v "^$" | sed -e "s/,$//" -e "s/.*\.\(.*\)\.\(.*\)$/\1.\2/" | sort | uniq -c)
        
        ## choosing top 2 sequences
        top_alleles=$(echo "$allele_counts" | sort -k1n,1 | tail -2)

        ## the percentage of each of the top 2 sequences relative to the total number of reads is calculated
        percentages=$(echo "$top_alleles" | awk -v total="$sum_counts" '\''{printf "%.2f%%\n", ($1 * 100 / total)}'\'')

        ## if the difference between the percentages of the top 2 alleles is greater than 50
        ## the higher percentage allele is classified as the typed allele, suggesting theres only one dominant alle
        ## otherwise both alleles are considered, suggesting there are two alleles
        allele1_percentage=$(echo "$percentages" | awk 'NR==1{print}')
        allele2_percentage=$(echo "$percentages" | awk 'NR==2{print}')
        percentage_difference=$(awk -v a="$allele2_percentage" -v b="$allele1_percentage" '\''BEGIN {printf "%.2f\n", a - b }'\'')
        percentage_difference=$(echo "$percentage_difference" | sed 's/,/./g')
        percentages=$(echo "$percentages" | sed 's/,/./g')

        if [ "$(awk -v pd="$percentage_difference" "BEGIN {printf (pd > 50) ? "1" : "0"}")" -eq 1 ]; then
            typed_allele=$(echo "$top_alleles" | tail -1)
        else
            typed_allele=$(echo "$top_alleles")
        fi

        ## writing the output
        echo "$header: $sequence" >> "$output_file"
        echo "Top 2 alleles:" >> "$output_file"
        echo "$top_alleles" >> "$output_file"
        echo "Total number of reads: $sum_counts" >> "$output_file"
        echo "Percentage of the top 2 alleles:" >> "$output_file"
        echo "$percentages" >> "$output_file"
        echo "Difference between the top 2 alleles: $percentage_difference%" >> "$output_file"
        echo "Typed alleles:" >> "$output_file"
        echo "$typed_allele" >> "$output_file"
        echo "----------------------" >> "$output_file"

    done < "$seq"
' _ "$output_dir" "$seq"